getAndWatch_stream_consumer.go

```diff
@@ -0,0 +1,203 @@
+	}
+	return wasAlreadyStopped
+}
+
+type getAndWatchConsumer struct {
+	endpoint   *streamEndpoint
+	streamName string
+	evtChan    chan *stream.GetAndWatchEvent
+	config     *ConsumerConfig
+	stopped    *int32
+}
+
+func (c *getAndWatchConsumer) streamEndpoint() *streamEndpoint {
+	return c.endpoint
+}
+
+func (c *getAndWatchConsumer) StreamName() string {
+	return c.streamName
+}
+
+func (c *getAndWatchConsumer) EvtChan() chan *stream.GetAndWatchEvent {
+	return c.evtChan
+}
+
+func (c *getAndWatchConsumer) Stop() bool {
+	return atomic.SwapInt32(c.stopped, 1) == 1
+}
+
+func (c *getAndWatchConsumer) isStopped() bool {
+	return atomic.LoadInt32(c.stopped) == 1
+}
+
+// Call this method to create a stream consumer
+// The service name is resolved via service discovery
+// Under the hood we make sure that only 1 subscription is done for a service, even if multiple streams are created on the same service
+func (g *Gaz) GetAndWatchStream(service, stream string, opts ...ConsumerConfigOpt) (GetAndWatchStreamConsumer, error) {
+	return g.createGetAndWatchConsumer([]string{SdPrefix + service}, stream, opts...)
+}
+
+func (g *Gaz) createGetAndWatchConsumer(endpoints []string, streamName string, opts ...ConsumerConfigOpt) (GetAndWatchStreamConsumer, error) {
+	r := g.streamConsumers
+	target := strings.Join(endpoints, ",")
+	r.Lock()
+	defer r.Unlock()
+	e, ok := r.endpointsByName[target]
+	if !ok {
+		var err error
+		Log.Debug("Creating stream endpoint", zap.String("target", target))
+		e, err = r.g.newStreamEndpoint(endpoints, g.streamEndpointOptions...)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error while creating stream endpoint for target %s", target)
+		}
+		r.endpointsByName[e.target] = e
+	}
+	sc := e.getAndWatch(streamName, opts...)
+	rc := registeredGetAndWatchConsumer{g: r.g, GetAndWatchStreamConsumer: sc}
+	consumers := r.endpointConsumers[e]
+	if consumers == nil {
+		consumers = make(map[StoppableStream]struct{})
+		r.endpointConsumers[e] = consumers
+	}
+	consumers[&rc] = struct{}{}
+	return &rc, nil
+}
+
+func (se *streamEndpoint) getAndWatch(streamName string, opts ...ConsumerConfigOpt) GetAndWatchStreamConsumer {
+	config := defaultConsumerConfig()
+	for _, opt := range opts {
+		opt(config)
+	}
+
+	ch := make(chan *stream.GetAndWatchEvent, config.BufferLen)
+	c := &getAndWatchConsumer{
+		endpoint:   se,
+		streamName: streamName,
+		evtChan:    ch,
+		config:     config,
+		stopped:    new(int32),
+	}
+
+	var monitoringHolder = consumerMonitoring(se.g, streamName, se.endpoints)
+
+	go func() {
+		se.reconnectGetAndWatchWhileNotStopped(c, streamName, config, monitoringHolder)
+		Log.Info("Stream closed", zap.String("stream", c.streamName))
+		close(c.evtChan)
+	}()
+	return c
+}
+
+func (se *streamEndpoint) reconnectGetAndWatchWhileNotStopped(c *getAndWatchConsumer, streamName string, config *ConsumerConfig, monitoringHolder consumerMonitoringHolder) {
+	for se.conn.GetState() != connectivity.Shutdown && !c.isStopped() {
```

Comment:
there is a gRPC option to wait for ready
I'm not sure it's a good idea to use it, I didn't read the doc carefully enough
https://github.com/grpc/grpc/blob/master/doc/wait-for-ready.md

https://gowalker.org/github.com/grpc/grpc-go#WaitForReady